### PR TITLE
Add a test to reproduce the bcast deadlock problem

### LINKED ISSUE #7093: pytest does not detect Open MPI's hang issue on MNIST
The test does not detect the GPU-direct hang issue of Open MPI

As described in #5740 and #7086 , some Open MPI versions have bugs that cause ChainerMN MNIST example to hang. However, the test suite of ChainerMN ( `communicator_test` in particular) does not detect the problem. 

Actually, in my experiments
* MNIST hangs
* `communicator_test` passes
* `gpudirect_bug.c` fails with an error in MPI_Init()

(`gpudirect_bug.c` is from https://github.com/open-mpi/ompi/issues/3972)

Thus, we need further investigation.

The goal of this issue is
* Find the root cause of MNIST hang.
  * We assume the root cause is Open MPI's issue 3972, but not sure
  * The problem may include environment-dependency.
* Detect the error in the test suite (i.e. communicator_test)

### PR DESCRIPTION
**Pull request by keisukefukuda**
This PR fixes #7093 by adding a test to reproduce deadlock in mpi4py's Bcast function.

### CODE REVIEW

File: tests/chainermn_tests/communicator_tests/test_mpi.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+                p = Popen(['ompi_info', '--all', '--parsable'], stdout=PIPE)
+                out, err = p.communicate()
+                if type(out) == bytes:
+                    out = out.decode('utf-8')
+                m = re.search(r'ompi:version:full:(\S+)', out)
+                version = m.group(1)
+
+                msg = "\n\n***** ERROR:" \
+                      "It looks like you are using " \
+                      "Open MPI version {}." \
```
Comment (belldandyxtq) [inline tests/chainermn_tests/communicator_tests/test_mpi.py]: Can you add a line break here?

File: tests/chainermn_tests/communicator_tests/test_mpi.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+                      "It is known that the following Open MPI versions " \
+                      "have a bug that cause MPI_Bcast() deadlock when " \
+                      "GPUDirect is used: " \
```
Comment (belldandyxtq) [inline tests/chainermn_tests/communicator_tests/test_mpi.py]: And here as well

**Comment by belldandyxtq** [comment]
I have tested the following MPI versions
```
2.1.6     succeeded  (no error)
3.0.0     succeeded (hang, got error message)
3.0.1     failed? (core dumped)
3.0.2     failed? (core dumped)
3.0.3     failed (core dumped, NOT SUPPOSED)
3.1.0     succeeded (hang, got error message)
3.1.3     succeeded  (no error)
3.1.4     succeeded  (no error)
```
Will add `3.0.4, 3.1.1, 3.1.2` results
Command:
```
mpienv use $VERSION
mpienv exec -np 2 -npernode 1 -hostfile host --mca orte_base_help_aggregate 0 pytest test_mpi.py -s
```

Environment:
```
platform linux -- Python 3.7.2, pytest-4.6.2, py-1.8.0, pluggy-0.12.0
```

File: tests/chainermn_tests/communicator_tests/test_mpi.py (hunk 1 above)
Comment (belldandyxtq) [inline tests/chainermn_tests/communicator_tests/test_mpi.py]: Can you add a line break here?

File: tests/chainermn_tests/communicator_tests/test_mpi.py (hunk 2 above)
Comment (belldandyxtq) [inline tests/chainermn_tests/communicator_tests/test_mpi.py]: line break here

File: tests/chainermn_tests/communicator_tests/test_mpi.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            chainer.cuda.get_device_from_id(self.device).use()
+        else:
+            self.device = -1
+
+        if self.communicator.size < 2:
+            pytest.skip('This test is for at least two processes')
+
+        self.queue = queue.Queue(maxsize=1)
+
+    def teardown(self):
```
Comment (belldandyxtq) [inline tests/chainermn_tests/communicator_tests/test_mpi.py]: How about using the `tearDown`?

File: tests/chainermn_tests/communicator_tests/test_mpi.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+
+        # This Bcast() cause deadlock if the underlying MPI has the bug.
+        th = _TimeoutThread(self.queue, self.communicator.rank)
+        th.start()
+        mpi_comm.Bcast(ptr, root=0)
+        mpi_comm.barrier()
+        self.queue.put(True)
+        assert True
+
+        self.teardown()
```
Comment (belldandyxtq) [inline tests/chainermn_tests/communicator_tests/test_mpi.py]: So we do not need to call it explicitly

File: tests/chainermn_tests/communicator_tests/test_mpi.py (hunk 3 above)
Comment (keisukefukuda) [reply tests/chainermn_tests/communicator_tests/test_mpi.py]: Fixed!

File: tests/chainermn_tests/communicator_tests/test_mpi.py (hunk 4 above)
Comment (keisukefukuda) [reply tests/chainermn_tests/communicator_tests/test_mpi.py]: Fixed by ↑

File: tests/chainermn_tests/communicator_tests/test_mpi.py (hunk 1 above)
Comment (keisukefukuda) [reply tests/chainermn_tests/communicator_tests/test_mpi.py]: Improved the error message so the message would be:
```
***** ERROR: It looks like you are using Open MPI version 3.0.0.
***** It is known that the following Open MPI versions have a bug
***** that cause MPI_Bcast() deadlock when GPUDirect is used:
*****  3.0.0, 3.0.1, 3.0.2, 3.1.0, 3.1.1, 3.1.2
```

**Comment by keisukefukuda** [comment]
I will update the MPI version list according to our internal & detailed tests.

**Comment by belldandyxtq** [comment]
pfnci, test this please

**Comment by belldandyxtq** [comment]
LGTM, 
The errors are irrelative to this PR.

**Pull request merged**